# Exploration

This Notebook is aimed at helping you apply in a structured way what you've learned about programming and working with data using pandas to explore the challenge dataset and contextualize the information we have with the business objective.

It's important to do so in a structured, organized way. Documenting your work properly and making it readable is essential to clearly communicate your results and keep track of all tasks you've done. 

## Import the dataset

Data files will be provided as "csv" files, which can be imported in pandas given the filepath to the pd.read_csv(FILE_PATH) function.

In [ ]:
import pandas as pd 

#df = pd.read_csv("Zambrano(1).csv") ### ADD YOUR FILE PATH HERE.
df = pd.read_csv("Data/Zambrano (1).csv")



pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
pd.set_option("display.width", 2000)
pd.set_option("display.max_colwidth", None)

df.head(5).T


,0,1,2,3,4
sucursal,Panem - Hospital Zambrano N,Panem - Hospital Zambrano N,Panem - Hospital Zambrano N,Panem - Hospital Zambrano N,Panem - Hospital Zambrano N
operating_date,2026-02-05,2026-02-05,2026-02-05,2026-02-05,2026-02-05
day_name,jueves,jueves,jueves,jueves,jueves
closing_time,2026-02-05 08:09:04.327000,2026-02-05 08:09:04.327000,2026-02-05 08:09:04.327000,2026-02-05 08:09:04.327000,2026-02-05 08:09:04.327000
captured_time,2026-02-05 08:05:09.710000,2026-02-05 08:05:18.727000,2026-02-05 08:05:40.727000,2026-02-05 08:05:41.490000,2026-02-05 08:05:42.210000
week_number,6,6,6,6,6
pdv_txn_id,179310,179310,179310,179310,179310
order_id,1,1,1,1,1
order_type,Para llevar,Para llevar,Para llevar,Para llevar,Para llevar
order_subtype,'-,'-,'-,'-,'-


To visualize what columns are present, we can use the .columns attribute

In [24]:
df.columns
df.shape

print(df.columns)
print(df.shape)

Index(['sucursal', 'operating_date', 'day_name', 'closing_time', 'captured_time', 'week_number', 'pdv_txn_id', 'order_id', 'order_type', 'order_subtype', 'table_number', 'party_size', 'server', 'terminal', 'capture_terminal', 'action', 'item', 'modifier', 'group_type', 'group', 'description', 'is_modifier', 'quantity', 'unit_price', 'unit_price_with_mods', 'cost_actual', 'cost_with_mods', 'cost_ideal', 'discount', 'subtotal_ticket', 'iva_ticket', 'ieps_ticket', 'total_ticket', 'subtotal_item', 'iva_item', 'ieps_item', 'total_item', 'subtotal_cortesia_cancel', 'iva_cortesia_cancel', 'ieps_cortesia_cancel', 'total_cortesia_cancel', 'subtotal_anulacion', 'iva_anulacion', 'ieps_anulacion', 'total_anulacion', 'clave_platillo'], dtype='str')
(4268, 46)


## Data Description

Briefly describe:
* What the dataset contains (in your own words) This dataset contains information about panem's Zambrano Branch, we can find information about the type order, the total item, top items, operation time, and other relevant details to understand the business.
* How many rows and columns? This dataset contains 4268 rows and 46 columns



In [25]:
from numpy import dtype
### YOUR CODE HERE


df.dtypes.to_frame("dtype")


,dtype
sucursal,str
operating_date,str
day_name,str
closing_time,str
captured_time,str
week_number,int64
pdv_txn_id,int64
order_id,int64
order_type,str
order_subtype,str


## Details

Answer the following questions:

1. What does one single row represent?
It represents a single transaction or sale event.
2. What does each column represent? 
    Which columns are categorical, numerical, datetime, or ID (identifier)? 

Categorical: sucursal, day_name, order_type, order_subtype, server, terminal, capture_terminal, action, item, modifier, group_type, group, description, clave_platillo, is_modifier
Datetime: operation_date, closing_time, captured_time
ID (identifier): pdv_txn_id, order_id, table_number
Numerical: week_number, party_size,quantity, unit_price, unit_price_with_mods, cost_actual, cost_with_mods, cost_ideal, discount, subtotal_ticket, iva_ticket, ieps_ticket, total_ticket, subtotal_item, iva_item, ieps_item, total_item, subtotal_cortesia_cancel, iva_cortesia_cancel, ieps_cortesia_cancel, total_cortesia_cancel, subtotal_anulacion, iva_anulacion, ieps_anulacion, total_anulacion



## Data Types and Conversions

Answer: 
1. 
Categorical: sucursal, day_name, order_type, order_subtype, server, terminal, capture_terminal, action, item, modifier, group_type, group, description, clave_platillo, is_modifier
Datetime: operation_date, closing_time, captured_time
ID (identifier): pdv_txn_id, order_id, table_number
Numerical: week_number, party_size,quantity, unit_price, unit_price_with_mods, cost_actual, cost_with_mods, cost_ideal, discount, subtotal_ticket, iva_ticket, ieps_ticket, total_ticket, subtotal_item, iva_item, ieps_item, total_item, subtotal_cortesia_cancel, iva_cortesia_cancel, ieps_cortesia_cancel, total_cortesia_cancel, subtotal_anulacion, iva_anulacion, ieps_anulacion, total_anulacion

Ensure that the columns are in the proper data type for their use, and convert them in case they aren't. Include text and code, showing your process and the final result of your work. 

In [26]:
df.duplicated().sum()

np.int64(0)

In [27]:
#does it have null values?
df.shape, df.isna().sum()

((4268, 46),
 sucursal                       0
 operating_date                 0
 day_name                       0
 closing_time                   0
 captured_time                  0
 week_number                    0
 pdv_txn_id                     0
 order_id                       0
 order_type                     0
 order_subtype                  0
 table_number                3496
 party_size                     0
 server                         0
 terminal                       0
 capture_terminal               3
 action                         0
 item                           0
 modifier                    3050
 group_type                     0
 group                          0
 description                    0
 is_modifier                    3
 quantity                       0
 unit_price                     3
 unit_price_with_mods           3
 cost_actual                    3
 cost_with_mods                 3
 cost_ideal                     3
 discount                       3
 

In [28]:
#Eliminamos valores no existentes y cambiamos el tipo de dato a string
df["modifier"] = df["modifier"].fillna("No modificacion").astype("string")

In [29]:
#Eliminamos los valores nulos y convertimos la columna a entero
df["table_number"] = df["table_number"].fillna(0).astype("int")

In [30]:
#Eliminamos los valores nulos y convertimos la columna a entero
df["capture_terminal"] = df["capture_terminal"].fillna("NULL").astype("str")

In [31]:
#Definir columnas

columnas_int = ["party_size","week_number", "pdv_txn_id", "order_id", "table_number"]
columnas_float = ["quantity", "unit_price", "unit_price_with_mods", "cost_actual", "cost_with_mods", "cost_ideal", "discount", "subtotal_ticket", "iva_ticket", "ieps_ticket", "total_ticket", "subtotal_item", "iva_item", "ieps_item", "total_item", "subtotal_cortesia_cancel", "iva_cortesia_cancel", "ieps_cortesia_cancel", "total_cortesia_cancel", "subtotal_anulacion", "iva_anulacion", "ieps_anulacion", "total_anulacion"]
columnas_datetime = ["operation_date", "closing_time", "captured_time"]
columnas_str = ["sucursal", "day_name", "order_type", "order_subtype", "server", "terminal", "capture_terminal", "action", "item", "modifier", "group_type", "group", "description", "clave_platillo"]
columnas_bool = ["is_modifier"]



In [32]:
df[columnas_float] = df[columnas_float].fillna(0).astype("float64")


In [33]:
for column in columnas_int:
  df[column] = df[column].astype("int")


In [34]:
df.shape

(4268, 46)

In [35]:
df["capture_terminal"].head(20)


0     SERVER1
1     SERVER1
2     SERVER1
3     SERVER1
4     SERVER1
5     SERVER1
6     SERVER1
7     SERVER1
8     SERVER1
9     SERVER1
10    SERVER1
11    SERVER1
12    SERVER1
13    SERVER1
14    SERVER1
15    SERVER1
16    SERVER1
17    SERVER1
18    SERVER1
19    SERVER1
Name: capture_terminal, dtype: str

In [36]:
df.shape, df.isna().sum()

((4268, 46),
 sucursal                    0
 operating_date              0
 day_name                    0
 closing_time                0
 captured_time               0
 week_number                 0
 pdv_txn_id                  0
 order_id                    0
 order_type                  0
 order_subtype               0
 table_number                0
 party_size                  0
 server                      0
 terminal                    0
 capture_terminal            0
 action                      0
 item                        0
 modifier                    0
 group_type                  0
 group                       0
 description                 0
 is_modifier                 3
 quantity                    0
 unit_price                  0
 unit_price_with_mods        0
 cost_actual                 0
 cost_with_mods              0
 cost_ideal                  0
 discount                    0
 subtotal_ticket             0
 iva_ticket                  0
 ieps_ticket              

## Exploration

Use different exploration methods such as .describe() (check the class activities and presentations) to inspect patterns in the data, and to spot potential inconsistencies. 

Look for values that don't make sense in a real-world context. 

In [41]:
df.describe()

,week_number,pdv_txn_id,order_id,table_number,party_size,quantity,unit_price,unit_price_with_mods,cost_actual,cost_with_mods,cost_ideal,discount,subtotal_ticket,iva_ticket,ieps_ticket,total_ticket,subtotal_item,iva_item,ieps_item,total_item,subtotal_cortesia_cancel,iva_cortesia_cancel,ieps_cortesia_cancel,total_cortesia_cancel,subtotal_anulacion,iva_anulacion,ieps_anulacion,total_anulacion
count,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.0000,4268.000000,4268.000000,4268.000000,4268.0,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.000000,4268.0,4268.0,4268.000000,4268.000000,4268.000000,4268.00000,4268.000000
mean,6.554124,179934.435333,80.883786,3.176898,1.062793,1.029522,59.550000,59.5500,10.367858,10.367858,16.502996,0.0,263.535426,32.530804,4.306755,300.372985,53.609855,6.388528,0.969219,60.967603,0.067685,0.0,0.0,0.073102,0.227786,0.024852,0.00446,0.258435
std,0.497120,363.287666,48.458706,7.789555,0.322273,0.326288,58.321566,60.6505,23.113053,23.113060,23.923157,0.0,187.929334,29.514934,5.829856,216.475785,52.231200,8.990937,1.999162,60.465466,3.721464,0.0,0.0,4.019206,4.728487,0.649878,0.17333,5.388274
min,6.000000,179310.000000,1.000000,0.000000,1.000000,1.000000,0.000000,0.0000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000,0.00000,0.000000
25%,6.000000,179620.750000,40.000000,0.000000,1.000000,1.000000,0.000000,0.0000,0.000000,0.000000,0.000000,0.0,125.000000,8.970000,0.000000,140.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000,0.00000,0.000000
50%,7.000000,179923.000000,78.000000,0.000000,1.000000,1.000000,45.000000,45.0000,1.517000,1.517000,7.320000,0.0,212.930000,26.210000,1.930000,244.000000,43.103400,0.000000,0.000000,45.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000,0.00000,0.000000
75%,7.000000,180261.000000,121.000000,0.000000,1.000000,1.000000,85.000000,87.0000,11.419025,11.419025,28.549500,0.0,347.160000,45.660000,6.670000,396.000000,77.586200,8.965500,0.000000,90.000000,0.000000,0.0,0.0,0.000000,0.000000,0.000000,0.00000,0.000000
max,7.000000,180563.000000,198.000000,29.000000,5.000000,10.000000,215.000000,236.0000,477.959600,477.959600,497.550000,0.0,993.100000,158.900000,42.230000,1152.000000,267.241400,42.758600,13.333300,310.000000,240.740000,0.0,0.0,260.000000,168.103400,26.896600,8.51850,195.000000


In [38]:
# rows that are duplicates of a previous row
columnas_duplicadas = df.duplicated()

# count duplicate rows
dup_count = columnas_duplicadas.sum()

# show duplicate rows
dups = df[columnas_duplicadas]
print(dup_count)
print(dups)


0
Empty DataFrame
Columns: [sucursal, operating_date, day_name, closing_time, captured_time, week_number, pdv_txn_id, order_id, order_type, order_subtype, table_number, party_size, server, terminal, capture_terminal, action, item, modifier, group_type, group, description, is_modifier, quantity, unit_price, unit_price_with_mods, cost_actual, cost_with_mods, cost_ideal, discount, subtotal_ticket, iva_ticket, ieps_ticket, total_ticket, subtotal_item, iva_item, ieps_item, total_item, subtotal_cortesia_cancel, iva_cortesia_cancel, ieps_cortesia_cancel, total_cortesia_cancel, subtotal_anulacion, iva_anulacion, ieps_anulacion, total_anulacion, clave_platillo]
Index: []


In [39]:
df["unit_price"].head(20)

0      55.0
1     155.0
2      55.0
3       0.0
4      55.0
5       0.0
6      60.0
7       0.0
8      55.0
9       0.0
10     55.0
11     90.0
12      0.0
13     65.0
14     26.0
15    125.0
16      0.0
17     40.0
18     65.0
19     77.0
Name: unit_price, dtype: float64

In [40]:
df["total_ticket"].min()
print(df["total_ticket"].min())
df["total_ticket"].max()
print(df["total_ticket"].max())

0.0
1152.0


## Prediction Shaping

When it comes to prediction, we are tryong to guess an unknown number using past examples. Each example must have the same format as what we want to predict in the future. 

For example:
* Using a house's number of rooms, total square footage, number of bedrooms and location to predict its price
* Predicting a plant's growth based on the amount of sunlight, water, and nutrients in the soil 

So before thinking about Machine Learning and AI, we need to answer: What does one thing we want to predict look like? What information would we use to predict what we want to predict, and what should that look like? 

Our training data should look just like our prediction data, except we already know the answer (the target variable's real value). For example, in the house price example, the question should be "how much did this house with 4 bedrooms, located in X neighborhood, with Y square footage sell for?" Your dataset should be able to answer that in a single row. 


Answer the questions:
1. What do we want to predict? Demand of the products to be sold, for efficiency and finding a balance between supply and demand, and to avoid overproduction and underproduction. 
2. What would a single row for prediction look like? 
a product of the store with the information and the Demand prediction
3. What do we need to do to the dataset to achieve this? 
we need to clean the data, remove duplicates, and remove the rows with missing values. 


## Open questions

Surely, you'll have a lot of questions about the dataset. Whenever you're exploring, it's a good idea to keep a log of unanswered questions, so that when you meet with the training partner again, you can make good use of their time with you. 

What questions do you have about the dataset that you can't yet answer? 
What information would you ask the company to better understand the dataset? 
Given the information in the dataset, what do you consider worth exploring in the later weeks? 

Are there known data issues or fields we should treat cautiously?


In [1]:
nuevo_leon = mexico_states[mexico_states["NOMGEO"] == "Nuevo León"]

fig, ax = plt.subplots(figsize=(8, 8))
nuevo_leon.plot(
    color="lightblue",
    edgecolor="black",
    ax=ax
)

ax.set_title("Nuevo León")
ax.axis("off")
plt.show()

NameError: name 'mexico_states' is not defined